### ISSUE TITLE
[feature request, idea] Vectorized base64 decoding / encoding

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#90560]
### 🚀 The feature, motivation and pitch

Discussed in context of scriptable base64 decoding here: https://github.com/pytorch/vision/issues/6878#issuecomment-1343439120

But then I thought that base64 can be implemented in a vectorized way - and probably some design could probably do batching (especially if inputs/outputs of same/padded size or nestedtensor/list)

The usecases are probably limited but could include some visualization utils that need to encode many images to base64 for further rendering/burning into html

**Comment by vadimkantorov** [comment]
Another useful encoding (e.g. in context of representing segmentation masks) is RLE: https://github.com/pytorch/pytorch/issues/32867#issuecomment-1072638742

**Comment by vadimkantorov** [comment]
If we have batched nonzero() returning nestedtensor/tensorlist, RLE encoding can also be done with existing ops. But might also be good to check the string format used by COCO internally to format to compact string the vals/counts `char* rleToString( const RLE *R ) {/* Similar to LEB128 but using 6 bits/char and ascii chars 48-111. */`

**Comment by vadimkantorov** [comment]
A timely :) recent paper on wavelet-compressed NerFs (using RLE as well): https://arxiv.org/abs/2212.09069

**Comment by vadimkantorov** [comment]
In the graphics world there are also all sort of formats for texture encoding/compression. And these can also be used in theory for compression of activation tensors (given they are simple/fast enough), so might be good to implement some of them for pytorch somehow

**Comment by vadimkantorov** [comment]
In the graphics world there are also all sort of formats for texture encoding/compression. And these can also be used in theory for compression of activation tensors (given they are simple/fast enough), so might be good to implement some of them for pytorch somehow

**Comment by vadimkantorov** [comment]
If any compression format can be supported in CUDA somehow (lzma? zstd?), then Apache Arrow )https://arrow.apache.org/docs/python/generated/pyarrow.compress.html) / HDF5 compressed data can save bandwidth on moving things from disk/RAM (by mmap or cached) to GPU memory

**Comment by vadimkantorov** [comment]
Related thread / paper on activation compression: https://twitter.com/arankomatsuzaki/status/1612260279873855490

**Comment by vadimkantorov** [comment]
btw RLE could be another sparse~ish tensor format

**Comment by vadimkantorov** [comment]
Related: 
- https://github.com/pytorch/vision/issues/8351

**Comment by vadimkantorov** [comment]
For encoding RLE in a COCO fashion I found this StackOverflow answer:
https://stackoverflow.com/a/76990451/445810
```python
import numpy as np

def binary_mask_to_rle_np(binary_mask):
    rle = {"counts": [], "size": list(binary_mask.shape)}

    flattened_mask = binary_mask.ravel(order="F")
    diff_arr = np.diff(flattened_mask)
    nonzero_indices = np.where(diff_arr != 0)[0] + 1
    lengths = np.diff(np.concatenate(([0], nonzero_indices, [len(flattened_mask)])))

    # note that the odd counts are always the numbers of zeros
    if flattened_mask[0] == 1:
        lengths = np.concatenate(([0], lengths))

    rle["counts"] = lengths.tolist()

    return rle

# For test
# def binary_mask_to_rle(binary_mask):
# ...
# M1 = np.array([1, 0, 0, 1])
# M2 = np.array([0, 1, 1, 0, 1, 1, 0])

# assert binary_mask_to_rle_np(M1)["counts"] == binary_mask_to_rle(M1)["counts"]
# assert binary_mask_to_rle_np(M2)["counts"] == binary_mask_to_rle(M2)["counts"]
```

For encoding masks, having something like this in torch.util would already be quite useful

For encoding other discrete things - probably as well

**Comment by cpuhrsch** [comment]
@vadimkantorov I also wrote a PyTorch batched version of RLE for SAM-fast a while ago
https://github.com/pytorch-labs/segment-anything-fast/blob/de861af9badd03e8b40f5e063d70754a3dc6b4f4/segment_anything_fast/utils/amg.py#L106-L141

It works well on GPU when combined with compile.

I never really bothered to split this out into its standalone util, so it might have sharp edges. Maybe this can serve as additional inspiration.

**Comment by vadimkantorov** [comment]
@msaroufim In https://pytorch.org/blog/accelerating-generative-ai-2/ it mentions `rle_dicts = [rle_dict_from_masks(m) for m in masks]`. I wonder if PyTorch now includes a CUDA kernel for masks<>RLE encoding?

**Comment by vadimkantorov** [comment]
For base64 decoding, here is my bite at it: https://gist.github.com/vadimkantorov/ea989c75f79961fe46182845b40d5f31

Two inefficiency issues with this:
- https://github.com/pytorch/pytorch/issues/61819#issuecomment-3089865206 could PyTorch support indexing with int8/uint8? in theory the error message says that indexing with "byte" tensor is possible, but in fact legacy uint8->bool casting prevents this
- Is there a way to express `(lhs | rhs).T.contiguous()` in a fused way, so that the output strides are transposed / the output matrix is stored in coumn-major? e.g. specify sth like `torch.bitwise_or(lhs, rhs, out = torch.empty_like(lhs).as_strided(lhs.stride[::-1])).T.view(-1)`

```python
# https://en.wikipedia.org/wiki/Base64

import torch

def decode_base64(base64_as_uint8_tensor):
    base64alphabet, base64pad = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/', '='
    decode = torch.zeros(256, dtype = torch.uint8, device = base64_as_uint8_tensor.device).put_(torch.tensor(list(map(ord, base64alphabet)), device = base64_as_uint8_tensor.device), torch.tensor(list(range(len(base64alphabet))), dtype = torch.uint8, device = base64_as_uint8_tensor.device))
    shifts = torch.tensor([[2], [4], [6], [4], [2], [0]], dtype = torch.uint8, device = base64_as_uint8_tensor.device)

    # still can't index with dtypes lower than int32 :( https://github.com/pytorch/pytorch/issues/61819#issuecomment-3089865206
    res = decode[base64_as_uint8_tensor.to(torch.int32)].view(-1, 4)
    resT = res.T.contiguous()
    lsh = resT[:-1] << shifts[:3]
    rsh = resT[ 1:] >> shifts[3:]
    res = (lsh | rsh).T.contiguous().view(-1)

    #res = torch.stack([(resT[0] << 2) | (resT[1] >> 4), (resT[1] << 4) | (resT[2] >> 2), (resT[2] << 6) | (resT[3] >> 0)], dim = -1).view(-1)

    unpad = (base64_as_uint8_tensor[-2:] == ord(base64pad)).sum()
    res = res[:res.shape[0] - unpad]

    return res

if __name__ == '__main__':
    inputstr1 = 'bGlnaHQgd29yay4='
    inputstr2 = 'bGlnaHQgd29yaw=='
    inputstr3 = 'bGlnaHQgd29y'

    for inputstr in [inputstr1, inputstr2, inputstr3]:
        base64_as_uint8_tensor = torch.tensor(list(map(ord, inputstr)), dtype = torch.uint8)
        decoded_as_uint8_tensor = decode_base64(base64_as_uint8_tensor)
        print(inputstr, [chr(b) for b in decoded_as_uint8_tensor.tolist()])
```
